File: dbxml/src/dbxml/XmlContainerConfig.cpp
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2002,2009 Oracle.  All rights reserved.
//
// $Id:$
// 

#include "dbxml/XmlContainerConfig.hpp"
#include "DbXmlInternal.hpp"
#include "ContainerConfig.hpp"

using namespace DbXml;

const char *XmlContainerConfig::NO_COMPRESSION = "NONE";
const char *XmlContainerConfig::DEFAULT_COMPRESSION = "DEFAULT";

XmlContainerConfig::XmlContainerConfig()
	: config_(new ContainerConfig())
{}

XmlContainerConfig::XmlContainerConfig(XmlContainer::ContainerType type,
	int mode, const char *name) 
	: config_(new ContainerConfig(type, mode, name))
{}

XmlContainerConfig::~XmlContainerConfig()
{
	delete config_;
}

XmlContainerConfig::XmlContainerConfig(const XmlContainerConfig &o)
	: config_(new ContainerConfig(*o))
{}

XmlContainerConfig::XmlContainerConfig(u_int32_t flags)
	: config_(new ContainerConfig(flags))
{}

XmlContainerConfig &XmlContainerConfig::operator=(const XmlContainerConfig &o)
{
	if (this != &o) {
		delete config_;
		config_ = new ContainerConfig(*o);
	}
	return *this;
}

const char *XmlContainerConfig::getCompressionName() const
{
	return config_->getCompressionName();
}
void XmlContainerConfig::setCompressionName(const char *name)
{
	config_->setCompressionName(name);
}

int XmlContainerConfig::getMode() const
{
	return config_->getMode();
}

void XmlContainerConfig::setMode(int mode)
{
	config_->setMode(mode);
}

XmlContainer::ContainerType XmlContainerConfig::getContainerType() const
{
	return config_->getContainerType();
}

void XmlContainerConfig::setContainerType(XmlContainer::ContainerType type)
{
	config_->setContainerType(type);
}

u_int32_t XmlContainerConfig::getPageSize() const
{
	return config_->getPageSize();
}

void XmlContainerConfig::setPageSize(u_int32_t pageSize)
{
	config_->setPageSize(pageSize);
}

u_int32_t XmlContainerConfig::getSequenceIncrement() const
{
	return config_->getSequenceIncrement();
}

void XmlContainerConfig::setSequenceIncrement(u_int32_t incr)
{
	config_->setSequenceIncrement(incr);
}

void XmlContainerConfig::setIndexNodes(ConfigState state)
{
	config_->setIndexNodes(state);
}

XmlContainerConfig::ConfigState XmlContainerConfig::getIndexNodes() const
{
	return config_->getIndexNodes();
}

void XmlContainerConfig::setChecksum(bool value)
{
	config_->setChecksum(value);
}

bool XmlContainerConfig::getChecksum() const
{
	return config_->getChecksum();
}

void XmlContainerConfig::setEncrypted(bool value)
{
	config_->setEncrypted(value);
}
bool XmlContainerConfig::getEncrypted() const
{
	return config_->getEncrypted();
}

void XmlContainerConfig::setAllowValidation(bool value)
{
	config_->setAllowValidation(value);
}
bool XmlContainerConfig::getAllowValidation() const
{
	return config_->getAllowValidation();
}

void XmlContainerConfig::setStatistics(ConfigState state)
{
	config_->setStatistics(state);
}

XmlContainerConfig::ConfigState XmlContainerConfig::getStatistics() const
{
	return config_->getStatistics();
}

void XmlContainerConfig::setTransactional(bool value)
{
	config_->setTransactional(value);
}
bool XmlContainerConfig::getTransactional() const
{
	return config_->getTransactional();
}

void XmlContainerConfig::setAllowCreate(bool value)
{
	config_->setAllowCreate(value);
}
bool XmlContainerConfig::getAllowCreate() const
{
	return config_->getAllowCreate();
}

void XmlContainerConfig::setExclusiveCreate(bool value)
{
	config_->setExclusiveCreate(value);
}
bool XmlContainerConfig::getExclusiveCreate() const
{
	return config_->getExclusiveCreate();
}

void XmlContainerConfig::setNoMMap(bool value)
{
	config_->setNoMMap(value);
}
bool XmlContainerConfig::getNoMMap() const
{
	return config_->getNoMMap();
}

void XmlContainerConfig::setReadOnly(bool value)
{
	config_->setReadOnly(value);
}
bool XmlContainerConfig::getReadOnly() const
{
	return config_->getReadOnly();
}

void XmlContainerConfig::setMultiversion(bool value)
{
	config_->setMultiversion(value);
}
bool XmlContainerConfig::getMultiversion() const
{
	return config_->getMultiversion();
}

void XmlContainerConfig::setReadUncommitted(bool value)
{
	config_->setReadUncommitted(value);
}
bool XmlContainerConfig::getReadUncommitted() const
{
	return config_->getReadUncommitted();
}

void XmlContainerConfig::setThreaded(bool value)
{
	config_->setThreaded(value);
}
bool XmlContainerConfig::getThreaded() const
{
	return config_->getThreaded();
}

void XmlContainerConfig::setTransactionNotDurable(bool value)
{
	config_->setTransactionNotDurable(value);
}
bool XmlContainerConfig::getTransactionNotDurable() const
{
	return config_->getTransactionNotDurable();
}

XmlContainerConfig::XmlContainerConfig(const char *name)
	: config_(new ContainerConfig())
{
	config_->setCompressionName(name);
}

XmlContainerConfig::operator u_int32_t() const
{
	return config_->getXmlFlags()|config_->getDbOpenFlags()|config_->getDbSetFlags()|config_->getSeqFlags();
}

XmlContainerConfig::XmlContainerConfig(const ContainerConfig &config)
	: config_(new ContainerConfig(config))
{}

XmlContainerConfig::XmlContainerConfig(int flags)
	: config_(new ContainerConfig(flags))
{}
	
XmlContainerConfig::XmlContainerConfig(XmlContainer::ContainerType type)
	: config_(new ContainerConfig())
{
	config_->setContainerType(type);
}

XmlContainerConfig DbXml::operator|(const XmlContainerConfig &a, const XmlContainerConfig &b)
{
	XmlContainerConfig config(b);
	(*config).setDbOpenFlags((*a).getDbOpenFlags()|(*b).getDbOpenFlags());
	(*config).setDbSetFlags((*a).getDbSetFlags()|(*b).getDbSetFlags());
	(*config).setSeqFlags((*a).getSeqFlags()|(*b).getSeqFlags());
	(*config).setXmlFlags((*a).getXmlFlags()|(*b).getXmlFlags());
	return config;
}

XmlContainerConfig DbXml::operator|(const char *name, const XmlContainerConfig &b)
{
	XmlContainerConfig config(b);
	config.setCompressionName(name);
	return config;
  
}

XmlContainerConfig DbXml::operator|(const XmlContainerConfig &a, const char *name)
{
	XmlContainerConfig config(a);
	config.setCompressionName(name);
	return config;
}

XmlContainerConfig DbXml::operator|(XmlContainer::ContainerType type, const XmlContainerConfig &b)
{
	XmlContainerConfig config(b);
	config.setContainerType(type);
	return config;
  
}

XmlContainerConfig DbXml::operator|(const XmlContainerConfig &a, XmlContainer::ContainerType type)
{
	XmlContainerConfig config(a);
	config.setContainerType(type);
	return config;
}
  
XmlContainerConfig DbXml::operator|(u_int32_t flags, const XmlContainerConfig &b)
{
	XmlContainerConfig config(flags);
	return config|b;
  
}

XmlContainerConfig DbXml::operator|(const XmlContainerConfig &a, u_int32_t flags)
{
	XmlContainerConfig config(flags);
	return config|a;
}

XmlContainerConfig DbXml::operator|(int flags, const XmlContainerConfig &b)
{
	XmlContainerConfig config(flags);
	return config|b;
  
}

XmlContainerConfig DbXml::operator|(const XmlContainerConfig &a, int flags)
{
	XmlContainerConfig config(flags);
	return config|a;
}


		

File: dbxml/src/java/com/sleepycat/dbxml/XmlContainerConfig.java

//See the file LICENSE for redistribution information.

//Copyright (c) 2002,2009 Oracle.  All rights reserved.




package com.sleepycat.dbxml;
import com.sleepycat.db.DatabaseConfig;
import com.sleepycat.db.internal.DbConstants;

//Adds configuration for:
//node indexing
//Use DatabaseConfig for:
//checksum, encryption, transactional configuration

public class XmlContainerConfig extends DatabaseConfig {

    public static final XmlContainerConfig DEFAULT = new XmlContainerConfig();
    public static final String DEFAULT_COMPRESSSION = "DEFAULT";
    public static final String NO_COMPRESSSION = "NONE";
    public static final int UseDefault = 0;
    public static final int On = 1;
    public static final int Off = 2;

    public XmlContainerConfig() {
	setMode(0);
    }

    public String getCompression() {
	return compression;
    }

    public XmlContainerConfig setCompression(String compressionName) {
	compression = compressionName;
	return this;
    }

    /** @deprecated use getIndexNodesFlag  */
    @Deprecated 
    public boolean getIndexNodes() {
	if (indexNodesFlag == Off)
	    return false;
	else if (indexNodesFlag == On)
	    return true;
	if (nodeContainer == XmlContainer.NodeContainer)
	    return true;
	else
	    return false;
    }

    public int getIndexNodesFlag() {
	return indexNodesFlag;
    }

    /** @deprecated use setIndexNodes(int) */
    @Deprecated 
    public XmlContainerConfig setIndexNodes(boolean value) {
	if(value)
	    indexNodesFlag = On;
	else
	    indexNodesFlag = Off;
	return this;
    }

    public XmlContainerConfig setIndexNodes(int value) {
	indexNodesFlag = value;
	return this;
    }

    /** @deprecated use getStatisticsEnabledFlag  */
    @Deprecated 
    public boolean getStatisticsEnabled() {
	if(statisticsFlag == Off)
	    return false;
	else
	    return true;
    }

    public int getStatisticsEnabledFlag() {
	return statisticsFlag;
    }

    /** @deprecated use setStatisticsEnabled(int) */
    @Deprecated 
    public XmlContainerConfig setStatisticsEnabled(boolean value) {
	if(value)
	    statisticsFlag = On;
	else
	    statisticsFlag = Off;
	return this;
    }

    public XmlContainerConfig setStatisticsEnabled(int value) {
	statisticsFlag = value;
	return this;
    }

    /** @deprecated use getContainerType 
     */
    @Deprecated 
    public boolean getNodeContainer() {
	if (nodeContainer == XmlContainer.NodeContainer)
	    return true;
	return false;
    }

    public int getContainerType() {
	return nodeContainer;
    }

    /** @deprecated use setContainerType
     */
    @Deprecated public XmlContainerConfig setNodeContainer(boolean value) {
	if (value)
	    nodeContainer = XmlContainer.NodeContainer;
	else
	    nodeContainer = XmlContainer.WholedocContainer;
	return this;
    }

    public XmlContainerConfig setContainerType(int value) {
	nodeContainer = value;
	return this;
    }

    public boolean getAllowValidation() {
	return allowValidation;
    }

    public XmlContainerConfig setAllowValidation(boolean value) {
	this.allowValidation = value;
	return this;
    }

    public void setSequenceIncrement(int incr) {
	sequenceIncr = incr;
    }

    public int getSequenceIncrement() {
	return sequenceIncr;
    }

    /* package */
    
    protected XmlContainerConfig(int[] data)
    {
    	setMode(data[0]);
    	setContainerType(data[1]);
    	setXmlFlags(data[2]);
    	setDbOpenFlags(data[3]);
    	setPageSize(data[4]);
    	setSequenceIncrement(data[5]);
    	setDbSetFlags(data[6]);
    	setSeqFlags(data[7]);
    }
    
    /* packs the XmlContainerConfig data into an array so
     * it can be easily passed to the C++.
     */
    protected int [] pack()
    {
	int [] data = new int[8];

	data[0] = (int)getMode(); //mode
	data[1] = nodeContainer; //container type
	data[2] = getXmlFlags(); //Flags beginning in DBXML_
	data[3] = getDbOpenFlags(); //Flags used to open a database
	data[4] = getPageSize(); //Page size
	data[5] = sequenceIncr; //Sequence increment
	data[6] = getDbSetFlags(); //Flags used for db->set_flags
	data[7] = getSeqFlags(); //Flags used for opening a sequence database
	return data;           
    }

    protected void setThreaded(boolean value)
    {
	isThreaded = value;
    }
    
    private void setXmlFlags(int xmlFlags)
    {
    	if ((xmlFlags & dbxml_javaConstants.DBXML_INDEX_NODES) != 0)
    		indexNodesFlag = On;
    	else if ((xmlFlags & dbxml_javaConstants.DBXML_NO_INDEX_NODES) != 0)
    		indexNodesFlag = Off;
    	else
    		indexNodesFlag = UseDefault;
    	if ((xmlFlags & dbxml_javaConstants.DBXML_STATISTICS) != 0)
    		statisticsFlag = On;
    	else if ((xmlFlags & dbxml_javaConstants.DBXML_NO_STATISTICS) != 0)
    		statisticsFlag = Off;
    	else
    		statisticsFlag = UseDefault;
    	setTransactional((xmlFlags & dbxml_javaConstants.DBXML_TRANSACTIONAL) != 0);
    	setAllowValidation((xmlFlags & dbxml_javaConstants.DBXML_ALLOW_VALIDATION) != 0);
    	setChecksum((xmlFlags & dbxml_javaConstants.DBXML_CHKSUM) != 0);
    	if ((xmlFlags & dbxml_javaConstants.DBXML_ENCRYPT) != 0)
    		setEncrypted("");
    	else
    		setEncrypted(null);
    }
    
    private void setDbOpenFlags(int dbFlags)
    {
    	setAllowCreate((dbFlags & DbConstants.DB_CREATE) != 0);
    	setExclusiveCreate((dbFlags & DbConstants.DB_EXCL) != 0);
    	setNoMMap((dbFlags & DbConstants.DB_NOMMAP) != 0);
    	setReadOnly((dbFlags & DbConstants.DB_RDONLY) != 0);
    	setMultiversion((dbFlags & DbConstants.DB_MULTIVERSION) != 0);
    	setReadUncommitted((dbFlags & DbConstants.DB_READ_UNCOMMITTED) != 0);
    	isThreaded = (dbFlags & DbConstants.DB_THREAD) != 0;
    }
    
    private void setDbSetFlags(int dbFlags)
    {
    	if ((dbFlags & dbxml_javaConstants.DBXML_ENCRYPT) != 0)
    		setEncrypted("");
    	else
    		setEncrypted(null);
    	setChecksum((dbFlags & dbxml_javaConstants.DBXML_CHKSUM) != 0);
    	setTransactionNotDurable((dbFlags & DbConstants.DB_TXN_NOT_DURABLE) != 0);
    }
    
    private void setSeqFlags(int dbFlags)
    {
    	setAllowCreate((dbFlags & DbConstants.DB_CREATE) != 0);
    	setExclusiveCreate((dbFlags & DbConstants.DB_EXCL) != 0);
    	isThreaded = (dbFlags & DbConstants.DB_THREAD) != 0;
    }

    private int getXmlFlags() {
	int flags = 0;
	// if indexNodesFlag isn't set, use default based on container type
	if (indexNodesFlag == On)
	    flags |= dbxml_javaConstants.DBXML_INDEX_NODES;
	else if (indexNodesFlag == Off)
	    flags |= dbxml_javaConstants.DBXML_NO_INDEX_NODES;
	if (statisticsFlag == On)
	    flags |= dbxml_javaConstants.DBXML_STATISTICS;
	else if (statisticsFlag == Off)
	    flags |= dbxml_javaConstants.DBXML_NO_STATISTICS;
	// Translations from DB
	if (getTransactional())
	    flags |= dbxml_javaConstants.DBXML_TRANSACTIONAL;
	if (getAllowValidation())
	    flags |= dbxml_javaConstants.DBXML_ALLOW_VALIDATION;
	if (getChecksum())
	    flags |= dbxml_javaConstants.DBXML_CHKSUM;
	if (getEncrypted())
	    flags |= dbxml_javaConstants.DBXML_ENCRYPT;
	return flags;
    }
    private int getDbOpenFlags() {
	int flags = 0;
	if (getAllowCreate())
	    flags |= DbConstants.DB_CREATE;
	if (getExclusiveCreate())
	    flags |= DbConstants.DB_EXCL;
	if (getNoMMap())
	    flags |= DbConstants.DB_NOMMAP;
	if (getReadOnly())
	    flags |= DbConstants.DB_RDONLY;
	if (getMultiversion())
	    flags |= DbConstants.DB_MULTIVERSION;
	if (getReadUncommitted())
	    flags |= DbConstants.DB_READ_UNCOMMITTED;
	if (isThreaded)
	    flags |= DbConstants.DB_THREAD;
	return flags;
    }
    
    private int getDbSetFlags() {
    	int flags = 0;
    	if (getChecksum())
    	    flags |= dbxml_javaConstants.DBXML_CHKSUM;
    	if (getEncrypted())
    	    flags |= dbxml_javaConstants.DBXML_ENCRYPT;
    	if (getTransactionNotDurable())
    	    flags |= DbConstants.DB_TXN_NOT_DURABLE;
    	return flags;
    }
    
    private int getSeqFlags() {
    	int flags = 0;
    	if (getAllowCreate())
    	    flags |= DbConstants.DB_CREATE;
    	if (getExclusiveCreate())
    	    flags |= DbConstants.DB_EXCL;
    	if (isThreaded)
    	    flags |= DbConstants.DB_THREAD;
    	return flags;
        }

    private String compression = DEFAULT_COMPRESSSION; 
    /* Flags */
    private int indexNodesFlag = UseDefault;
    private int statisticsFlag = UseDefault;
    private boolean isThreaded = false;
    private int nodeContainer = XmlContainer.NodeContainer;
    private boolean allowValidation = false;
    private int sequenceIncr = 5;
}
